americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+        The title to be placed on the graph.
+    ylims : list
+        The limits for the y-axis.
+    ylabel : str
+        The label text for the y-axis. Every time you leave off appropriate
+        labels and units, a science grad student grading lab reports cries
+        another bitter tear into their bottle of craft beer.
+
+    Returns
+    -------
+    ax : axes
+        A matplotlib axes containing the plotted data
+
+    """
+
+    grouped = meta.groupby(group)
+
+    # Sets up the plotting order
+    if order is None:
+        order = grouped.groups.keys()
+
+    # Gets the data vectors
+    vecs = [grouped.get_group(g)[cat].values for g in order]
+
+    # Formats the axis, if not already done
+    if 'xticklabels' not in boxplot_props:
+        boxplot_props['xticklabels'] = [g.split('(')[0] for g in order]
+    if 'show_xticks' not in boxplot_props:
+        boxplot_props['show_xticks'] = False
+    if 'show_ygrid' not in boxplot_props:
+        boxplot_props['show_ygrid'] = True
+
+    # Calculates the p value
+    h, p = kruskal(*vecs)
+
+    # Sets the boxplot properties
+    ax = boxplot(vecs=vecs, ax=ax, p_value=p, **boxplot_props)
+
+    return ax
+
+
+def post_hoc_pandas(meta, group, cat, order=None, correct=None,
+    show_stats=True):
+    """Preforms an post-hoc comparison between two groups
+
+    Parameters
+    ----------
+    meta : pandas DataFrame
+        the metadata object for the data
+    group : str
+        the metadata category being interograted
+    cat : str
+        the name of the column with the result
+    order : None, list, optional
+        Default is None. The order of groups in the category.
+    correct : None, str, optional
+        Method for multiple hypothesis correction using
+        `statsmodels.sandbox.stats.multicomp.multipletests`. Methods you're
+        likely to use are `bonferroni` and `fdr_bh`.
+    show_stats : bool, optional
+        When `show_stats` is True, a summary of each group will be displayed
+        along with the p values. 
+
+    Returns
+    -------
+    post_hoc : dataframe
+        `post_hoc` summarizes the results of the post-hoc test. It includes
+        statitics about each distribution, as well as the comparison matrix
+        of p-values.
+
+    """
+
+    # Groups the data
+    grouped = meta.groupby(group)
+
+    # Gets the order
+    if order is None:
+        order = grouped.groups.keys()
+
+    # Prealocates an output frame
+    if show_stats:
+        stats = pd.DataFrame({'Counts': grouped[cat].count(),
+                              'Mean':  grouped[cat].mean(),
+                              'Stdv': grouped[cat].std(),
+                              'Median': grouped[cat].median()})
+
+    # Preforms ad-hoc comparisons
+    comparison = {}
+
+    for pos, g1_name in enumerate(order[:-1]):
+        g1_data = grouped.get_group(g1_name)[cat]
+        compare = []
+        index = []
+        for id2, g2_name in enumerate(order):
```

Comment:
This could be done fancier (non-blocking comment) using numpy. something like:

```python
mat = arange(1, len(order)) + arange(1, len(order))[:, newaxis]
mat -= tril(c)
for g1_pos, g2_pos in (mat > 0).nonzero():
    # do stuff
```

I think that works. Basically, the pattern here is that you want to produce some pairwise matrix from a vector. In this case, your end result is a vector, but if I'm looking at this right, its effectively a flattened version of the upper triangle. So what you really care about are getting the corresponding index positions you care about in that matrix (which is what the initial voodoo is doing). You can then index directly into `order`. The downside is this doesn't account for the `nan`, though that could be accounted for in the matrix (e.g., construct a matrix of `nan`, populate by the the imputed indices, and the serialize). 

These type of bulk operations are almost always significantly faster if done in numpy than in python too.